include/appbase/execution_priority_queue.hpp

```diff
@@ -16,48 +16,91 @@ struct priority {
    static constexpr int highest     = std::numeric_limits<int>::max();
 };
 
+// the index of the default queue
+constexpr static int default_queue = 0;
+
 class execution_priority_queue : public boost::asio::execution_context
 {
 public:
+   execution_priority_queue()
+   {
+      // add the default queue
+      multi_queue_add_queue();
+   }
 
    template <typename Function>
-   void add(int priority, Function function)
+   void add(int priority, int index, Function function)
    {
       std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(priority, --order_, std::move(function)));
 
-      handlers_.push(std::move(handler));
+      handlers_queue_[index]->push(std::move(handler));
    }
 
    void clear()
    {
-      handlers_ = prio_queue();
+      for (auto& q: handlers_queue_)
+         q = std::make_unique<prio_queue>();
    }
    
    void execute_all()
    {
-      while (!handlers_.empty()) {
-         handlers_.top()->execute();
-         handlers_.pop();
+      while (!handlers_queue_[default_queue]->empty()) {
+         handlers_queue_[default_queue]->top()->execute();
+         handlers_queue_[default_queue]->pop();
       }
    }
 
    bool execute_highest()
    {
-      if( !handlers_.empty() ) {
-         handlers_.top()->execute();
-         handlers_.pop();
+      if( !handlers_queue_[default_queue]->empty() ) {
+         handlers_queue_[default_queue]->top()->execute();
+         handlers_queue_[default_queue]->pop();
       }
 
-      return !handlers_.empty();
+      return !handlers_queue_[default_queue]->empty();
+   }
+
+   size_t size() { return handlers_queue_[default_queue]->size(); }
+
+   // multi-queue support
+
+   void multi_queue_register_next_handler_func(std::function<std::tuple<bool, int, bool>()> f) {
+      next_handler_func_ = f;
+   }
+
+   int multi_queue_add_queue() {
+      handlers_queue_.push_back( std::make_unique<prio_queue>() );
+      return handlers_queue_.size() - 1; // the index of the queue just added
+   }
+
+   int multi_queue_size(int i) {
+      return handlers_queue_[i]->size();
+   }
+
+   int multi_queue_empty(int i) {
+      return handlers_queue_[i]->empty();
+   }
+
+   bool multi_queue_less_than(int i, int j) {
+      return ( *handlers_queue_[i]->top() < *handlers_queue_[j]->top() );
    }
 
-   size_t size() { return handlers_.size(); }
+   bool multi_queue_execute_highest()
+   {
+      auto [has_handler, i, more] = next_handler_func_();
```

Comment:
I'm concerned about the overhead of `std::function` here vs lambda. Original implementation avoids use of `std::function`. Performance here can be very important.

Reply:
Thanks. I was thinking about the type safety of `next_handler_func_`. Will change to use lambda,